contracts/ResilientOracle.sol

```diff
@@ -329,8 +324,58 @@ contract ResilientOracle is PausableUpgradeable, AccessControlledV8, ResilientOr
         enabled = tokenConfigs[asset].enableFlagsForOracles[uint256(role)];
     }
 
+    /**
+     * @notice Updates the pivot oracle price. Currently using TWAP
+     * @dev Cache the asset price and return if already cached
+     * @param asset asset address
+     */
+    function _updateAssetPrice(address asset) internal {
+        if (_readCachedPrice(asset) != 0) {
+            return;
+        }
+
+        (address pivotOracle, bool pivotOracleEnabled) = getOracle(asset, OracleRole.PIVOT);
+        if (pivotOracle != address(0) && pivotOracleEnabled) {
+            //if pivot oracle is not TwapOracle it will revert so we need to catch the revert
+            try TwapInterface(pivotOracle).updateTwap(asset) {} catch {}
+        }
+
+        uint256 price = _getPrice(asset);
+        _cachePrice(asset, price);
+    }
+
+    /**
+     * @notice Cache the asset price into transient storage
+     * @param key address of the asset
+     * @param value asset price
+     */
+    function _cachePrice(address key, uint256 value) internal {
+        bytes32 slot = keccak256(abi.encode(CACHE_SLOT, key));
+        assembly ("memory-safe") {
+            tstore(slot, value)
+        }
+    }
+
+    /**
+     * @notice Read cached price from transient storage
+     * @param key address of the asset
+     * @return value cached asset price
+     */
+    function _readCachedPrice(address key) internal view returns (uint256 value) {
+        bytes32 slot = keccak256(abi.encode(CACHE_SLOT, key));
+        assembly ("memory-safe") {
+            value := tload(slot)
+        }
+    }
```

Comment:
We could extract these functions to an external file, to reuse it in other oracles caching prices